File: alds1/9a_complete_binary_tree.c
#define dbg(fmt,...) fprintf(stderr,fmt,__VA_ARGS__)
#define dpri(x) dbg(#x ": %d\n", x)
#define dprs(x) dbg(#x ": %s\n", x)
#include <stdio.h>
typedef long long ll;
const int MYINF = 1e9+7;
typedef int bool;
const bool true = 1;
const bool false = 0;

#define MAX_H 264
#define parent(i) ((i)/2)
#define left(i) ((i)*2)
#define right(i) ((i)*2+1)

int H;
int heap[MAX_H];
#define len heap[0]

void insert(int x) {
	heap[++len] = x;
}

void print_node(int i) {
	printf("node %d: key = %d, ", i, heap[i]);
	if (0 < parent(i)) printf("parent key = %d, ", heap[parent(i)]);
	if (left(i) <= len) printf("left key = %d, ", heap[left(i)]);
	if (right(i) <= len) printf("right key = %d, ", heap[right(i)]);
	puts("");
}

int main(int argc, char **argv)
{
	int i, x;
	scanf("%d", &H);
	for (i = 0; i < H; i++) {
		scanf("%d", &x);
		insert(x);
	}

	for (i = 1; i <= H; i++) {
		print_node(i);
	}

	return 0;
}

File: alds1/9a_complete_binary_tree.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"unicode"
)

type Heap struct {
	a []int
}

func NewHeap() *Heap {
	return &Heap{make([]int, 1)}
}

func parent(i int) int {
	return i / 2
}

func left(i int) int {
	return i * 2
}

func right(i int) int {
	return i*2 + 1
}

func (h *Heap) Push(k int) {
	h.a = append(h.a, k)
}

func (h *Heap) Print() {
	for i := 1; i < len(h.a); i++ {
		printf("node %d: key = %d, ", i, h.a[i])
		p := parent(i)
		l := left(i)
		r := right(i)
		if p > 0 {
			printf("parent key = %d, ", h.a[p])
		}
		if l < len(h.a) {
			printf("left key = %d, ", h.a[l])
		}
		if r < len(h.a) {
			printf("right key = %d, ", h.a[r])
		}
		println()
	}
}

func main() {
	defer stdout.Flush()
	n := readInt()
	h := NewHeap()
	for i := 0; i < n; i++ {
		k := readInt()
		h.Push(k)
	}
	h.Print()
}

var (
	stdin  *bufio.Reader
	stdout *bufio.Writer
)

func init() {
	stdin = bufio.NewReader(os.Stdin)
	stdout = bufio.NewWriter(os.Stdout)
}

func scanf(f string, args ...interface{}) (int, error) {
	n, err := fmt.Fscanf(stdin, f, args...)
	for {
		r, _, err := stdin.ReadRune()
		if err != nil || !unicode.IsSpace(r) {
			stdin.UnreadRune()
			break
		}
	}
	return n, err
}

func readInt() int {
	var n int
	scanf("%d", &n)
	return n
}

func readIntSlice(n int) []int {
	a := make([]int, n)
	for i := 0; i < n; i++ {
		scanf("%d", &a[i])
	}
	return a
}

func readLengthAndSlice() []int {
	n := readInt()
	return readIntSlice(n)
}

func readString() string {
	var s string
	scanf("%s", &s)
	return s
}

func printf(f string, args ...interface{}) (int, error) {
	return fmt.Fprintf(stdout, f, args...)
}

func println(args ...interface{}) (int, error) {
	return fmt.Fprintln(stdout, args...)
}
